File: trade-app/src/api/billing.ts
import { api } from "./client";

export type PlanCode = "pro" | "elite";

export type Subscription = {
  plan_code: string;
  plan_name: string;
  status: string;
  predictions_per_day: number;
  current_period_end: string | null;
  cancel_url: string | null;
};

export async function fetchSubscription(): Promise<Subscription> {
  const { data } = await api.get<Subscription>("/billing/subscription");
  return data;
}

export async function startCheckout(plan: PlanCode): Promise<string> {
  const { data } = await api.post<{ checkout_url: string; plan_code: PlanCode }>(
    "/billing/checkout",
    { plan_code: plan },
  );
  return data.checkout_url;
}

export async function openPortal(): Promise<string> {
  const { data } = await api.post<{ portal_url: string }>("/billing/portal");
  return data.portal_url;
}

File: backend/app/schemas/billing.py
"""Pydantic request/response shapes for /billing/*."""
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


PlanCode = Literal["pro", "elite"]


class CheckoutRequest(BaseModel):
    plan_code: PlanCode = Field(..., description="Which paid plan to subscribe to")


class CheckoutResponse(BaseModel):
    checkout_url: str
    plan_code: PlanCode


class PortalResponse(BaseModel):
    portal_url: str


class SubscriptionResponse(BaseModel):
    """What the frontend renders in the 'Current Plan' card."""
    plan_code: str
    plan_name: str
    status: str  # active, past_due, canceled, free
    predictions_per_day: int
    current_period_end: Optional[datetime] = None
    cancel_url: Optional[str] = None  # portal link when user has a real sub


class WebhookAck(BaseModel):
    received: bool = True
    event_type: Optional[str] = None
